backend/mem/file.go

```diff
@@ -0,0 +1,497 @@
+		return closeErr
+	}
+
+	_, serr4 := f.Seek(0, 0)
+	return serr4
+}
+
+/*
+ MoveToLocation moves the receiver file to the passed in location. It does so by
+creating a copy of 'f' in "location".  'f' is subsequently  deleted
+*/
+func (f *File) MoveToLocation(location vfs.Location) (vfs.File, error) {
+
+	if ex, eerr := f.Exists(); !ex {
+		if eerr == nil {
+			_, werr := f.Write(make([]byte, 0))
+			if werr != nil {
+				return nil, werr
+			}
+		} else {
+			return nil, doesNotExist()
+		}
+
+	}
+
+	testPath := path.Join(location.Path(), f.Name()) //this is a potential path to a file that can be fed into the objMap portion of fsMap
+	loc := location.(*Location)
+	mapRef := &loc.fileSystem.fsMap //mapRef just makes it easier to refer to "loc.fileSystem.fsMap"
+	vol := loc.Volume()
+	if _, ok := (*mapRef)[vol]; ok { //this checks if the specified volume has any keys
+		//this block checks if the file already exists at location, if it does, deletes it and inserts the file we have
+		if _, ok2 := (*mapRef)[vol][testPath]; ok2 { //if the file already exists at that location
+			file := (*mapRef)[vol][testPath].i.(*File)
+			cerr := f.CopyToFile(file)
+			if cerr != nil {
+				return nil, cerr
+			}
+
+			derr := f.Delete()
+			if derr != nil {
+				return nil, derr
+			}
+			return file, nil
+		}
+	}
+	// if the file doesn't yet exist at the location, create it there
+	newFile, nerr := location.NewFile(f.Name()) //creating the file in the desired location
+	if nerr != nil {
+		return nil, nerr
+	}
+
+	_, werr := newFile.Write(make([]byte, 0)) //initialize the file
+	if werr != nil {
+		return nil, werr
+	}
+	cerr := f.CopyToFile(newFile) //copying over the data
+	if cerr != nil {
+		return nil, cerr
+	}
+
+	derr := f.Delete()
+	if derr != nil {
+		return nil, derr
+	}
+	return newFile, nil
+}
+
+/*
+MoveToFile creates a newFile, and moves it to "file".
+If names are same, "file" is deleted and newFile takes its place.
+The receiver is always deleted (since it's being "moved")
+*/
+
+func (f *File) MoveToFile(file vfs.File) error {
+
+	if f.Name() == file.Name() {
+
+		loc := file.Location()
+
+		derr := file.Delete()
+		if derr != nil {
+			return derr
+		}
+
+		newFile, nerr := loc.NewFile(f.Name())
+		if nerr != nil {
+			return nerr
+		}
+
+		_, werr := newFile.Write(make([]byte, 0))
+		if werr != nil {
+			return werr
+		}
+		copyErr := f.CopyToFile(newFile)
+		if copyErr != nil {
+			return copyErr
+		}
+
+		return f.Delete()
+	}
+
+	newFile, nerr2 := file.Location().NewFile(f.Name())
+	if nerr2 != nil {
+		return nerr2
+	}
+	_, werr := newFile.Write(make([]byte, 0))
+	if werr != nil {
+		return werr
+	}
+	copyErr := f.CopyToFile(newFile)
+	if copyErr != nil {
+		return copyErr
+	}
+
+	return f.Delete()
+
+}
+
+/*
+Delete removes the file from the fs. Sets it path in the fsMap to nil,
+ and also nils the file's members
+*/
+func (f *File) Delete() error {
+
+	if ex, _ := f.Exists(); !ex {
+		return doesNotExist()
+	}
+
+	if f == nil {
```

Comment:
Should this conditional come before the prior Exists check? ie: it is meaningless to take hardly any action on a nil file pointer, so this check should come before a possible nil.Exists(). I'm not 100% sure of this, if you have a reason let me know.